# Week 3: Data Cleaning - Guided Exercise

**Duration:** ~30 minutes  
**Dataset:** Education Statistics from the Colombian Ministry of Education (datos.gov.co)  
**Rows:** 482 (dirty) | **Columns:** 37  
**Goal:** Clean this dataset by fixing 5 types of data quality issues  

---

**How to use this notebook:**

1. Read each explanation cell carefully
2. Run the code cell below it (Shift + Enter)
3. Read the "What just happened" follow-up
4. Answer the questions or complete the "Your Turn" exercises

Think of this entire process as doing laundry: we need to sort, wash, dry, and fold our data before it is ready to wear.

## Setup

We start by importing our tools and loading the CSV file. We also make a copy of the original data so we can compare before and after at the end.

**Run the cell below** to load the dataset.

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('data/educacion_estadisticas.csv')

# Keep a copy of the original so we can compare at the end
df_original = df.copy()

print(f"Dataset loaded: {df.shape[0]} rows, {df.shape[1]} columns")

Dataset loaded: 482 rows, 37 columns


The dataset has 482 rows and 37 columns. Each row represents one Colombian department in one year (2011-2024), with education indicators like enrollment rates, dropout rates, and coverage.

---

## Section 1: Data Inspection

Before cleaning anything, we need to understand what we are working with. Think of this as **opening the laundry bag** and checking what is inside before turning on the washing machine. You would not throw everything in without looking first.

This is the **inspection ritual**: a set of commands you run at the start of every data project.

### 1.1 Shape and Column Names

We use `df.shape` to see how many rows and columns exist, and `df.columns.tolist()` to see the names of all columns.

**Run the cell below.**

In [2]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")
print()
print("Column names:")
print(df.columns.tolist())

Rows: 482
Columns: 37

Column names:
['ano', 'c_digo_departamento', 'departamento', 'poblacion_5_16', 'tasa_matriculacion_5_16', 'cobertura_neta', 'cobertura_neta_transicion', 'cobertura_neta_primaria', 'cobertura_neta_secundaria', 'cobertura_neta_media', 'cobertura_bruta', 'cobertura_bruta_transicion', 'cobertura_bruta_primaria', 'cobertura_bruta_secundaria', 'cobertura_bruta_media', 'tamano_promedio_grupo', 'sedes_conectadas_a_internet', 'desercion', 'desercion_transicion', 'desercion_primaria', 'desercion_secundaria', 'desercion_media', 'aprobacion', 'aprobacion_transicion', 'aprobacion_primaria', 'aprobacion_secundaria', 'aprobacion_media', 'reprobacion', 'reprobacion_transicion', 'reprobacion_primaria', 'reprobacion_secundaria', 'reprobacion_media', 'repitencia', 'repitencia_transicion', 'repitencia_primaria', 'repitencia_secundaria', 'repitencia_media']


**What to notice:** The column names are in Spanish (as expected from a Colombian government dataset). Key columns include:
- `ano` = year
- `departamento` = department (geographic region)
- `poblacion_5_16` = population aged 5-16
- `desercion` = dropout rate
- `cobertura_neta` = net enrollment coverage
- `aprobacion` / `reprobacion` = approval / failure rates

### 1.2 First Rows

Looking at the actual data helps us spot problems that column names alone cannot reveal. `df.head()` shows the first 5 rows.

**Run the cell below.**

In [3]:
df.head()

ano  c_digo_departamento departamento poblacion_5_16  \
0  2023.0                   68    Santander         405256   
1  2024.0                   86     Putumayo          81100   
2  2016.0                   52     Nariño          394,574   
3  2024.0                   50         Meta         221911   
4  2020.0                    5    antioquia        1151149   

   tasa_matriculacion_5_16  cobertura_neta  cobertura_neta_transicion  \
0                    95.80           95.64                      71.27   
1                    80.03           80.01                      50.77   
2                    71.10             NaN                      38.56   
3                    90.34           90.29                      65.69   
4                    93.59           93.43                      70.25   

   cobertura_neta_primaria  cobertura_neta_secundaria  cobertura_neta_media  \
0                    92.98                      83.40                 56.02   
1                    78.12                      66.49                 38.12   
2                    68.32                      56.56                 29.80   
3                    89.17                      76.00                 48.10   
4                    88.26                      83.38                 47.45   

   ...  reprobacion  reprobacion_transicion  reprobacion_primaria  \
0  ...         7.65                    2.95                 91.82   
1  ...         6.07                    3.02                 90.68   
2  ...         1.89                    0.04                 97.71   
3  ...         6.84                    2.85                 91.18   
4  ...          NaN                    0.79                 92.19   

   reprobacion_secundaria  reprobacion_media  repitencia  \
0                   12.95               6.92        7.99   
1                   10.04               4.82        8.50   
2                    3.07               1.92        2.64   
3                   11.36               5.61        9.27   
4                   11.46               7.12        6.48   

   repitencia_transicion  repitencia_primaria  repitencia_secundaria  \
0                   2.95                 7.32                  11.34   
1                   3.02                 7.56                  12.33   
2                   0.04                 2.36                   3.79   
3                   2.85                 9.27                  12.53   
4                   0.79                 5.92                   9.37   

   repitencia_media  
0              4.07  
1              4.57  
2              1.34  
3              3.49  
4              3.17  

[5 rows x 37 columns]

### 1.3 Data Types

Every column has a data type that tells pandas how to store and process the values. We use `df.dtypes` to see them.

**Look for suspicious types:** a year stored as `float64` instead of `int64`, or a population number stored as `object` (text).

**Run the cell below.**

In [4]:
df.dtypes

ano                            float64
c_digo_departamento              int64
departamento                       str
poblacion_5_16                     str
tasa_matriculacion_5_16        float64
cobertura_neta                 float64
cobertura_neta_transicion      float64
cobertura_neta_primaria        float64
cobertura_neta_secundaria      float64
cobertura_neta_media           float64
cobertura_bruta                float64
cobertura_bruta_transicion     float64
cobertura_bruta_primaria       float64
cobertura_bruta_secundaria     float64
cobertura_bruta_media          float64
tamano_promedio_grupo          float64
sedes_conectadas_a_internet    float64
desercion                      float64
desercion_transicion           float64
desercion_primaria             float64
desercion_secundaria           float64
desercion_media                float64
aprobacion                     float64
aprobacion_transicion          float64
aprobacion_primaria            float64
aprobacion_secundaria    

**What just happened:** pandas tells us the type of each column.

**Key findings:**
- `ano` is `float64` (decimal) but years should be whole numbers (`int64`). Years like 2023.0 look odd.
- `poblacion_5_16` is `object` (text) even though it should be a number. This means some values contain characters that prevent pandas from reading them as numbers.
- Most rate columns are `float64`, which is correct for percentages.

### 1.4 Missing Values

`isnull().sum()` counts how many NaN (missing) values each column has. We sort the result so the worst offenders appear first.

**Run the cell below.**

In [5]:
missing = df.isnull().sum()
missing_only = missing[missing > 0].sort_values(ascending=False)

print(f"Columns with missing values: {len(missing_only)} of {len(df.columns)}")
print(f"Total missing cells: {missing.sum()} of {df.shape[0] * df.shape[1]:,}")
print()
print(missing_only)

Columns with missing values: 12 of 37
Total missing cells: 877 of 17,834

sedes_conectadas_a_internet    240
tamano_promedio_grupo          240
aprobacion                      55
reprobacion                     53
desercion_primaria              50
poblacion_5_16                  50
cobertura_neta_primaria         50
desercion                       47
desercion_secundaria            40
cobertura_neta                  40
departamento                    11
desercion_transicion             1
dtype: int64


### 1.5 Statistical Summary

`df.describe()` gives us count, mean, std, min, max, and percentiles for every numeric column. This helps spot outliers and confirms what we learned from `isnull()`.

**Run the cell below.**

In [6]:
df.describe().round(2)

ano  c_digo_departamento  tasa_matriculacion_5_16  cobertura_neta  \
count   482.00               482.00                   482.00          442.00   
mean   2017.47                51.96                    85.64           85.62   
std       4.02                30.89                    10.36           11.56   
min    2011.00                 5.00                    52.33           -9.26   
25%    2014.00                20.00                    80.01           80.46   
50%    2017.00                52.00                    87.53           88.05   
75%    2021.00                81.00                    93.46           93.43   
max    2024.00                99.00                   103.94          108.88   

       cobertura_neta_transicion  cobertura_neta_primaria  \
count                     482.00                   432.00   
mean                       57.72                    84.35   
std                        11.46                    10.28   
min                        19.22                    51.12   
25%                        51.41                    78.32   
50%                        58.82                    86.12   
75%                        66.16                    91.46   
max                        88.28                   135.06   

       cobertura_neta_secundaria  cobertura_neta_media  cobertura_bruta  \
count                     482.00                482.00           482.00   
mean                       68.22                 38.83            98.29   
std                        15.46                 13.08            11.29   
min                        -5.80                  6.79            59.79   
25%                        60.99                 31.18            93.32   
50%                        72.52                 41.07           100.46   
75%                        78.81                 48.08           105.92   
max                        91.09                113.09           123.04   

       cobertura_bruta_transicion  ...  reprobacion  reprobacion_transicion  \
count                      482.00  ...       429.00                  482.00   
mean                        89.94  ...         6.10                    1.45   
std                         17.03  ...         7.38                    2.34   
min                         32.03  ...       -14.42                    0.00   
25%                         81.09  ...         3.24                    0.22   
50%                         89.60  ...         6.19                    0.57   
75%                         97.88  ...         7.66                    1.54   
max                        144.97  ...       137.39                   17.01   

       reprobacion_primaria  reprobacion_secundaria  reprobacion_media  \
count                482.00                  482.00             482.00   
mean                  91.62                    8.24               5.20   
std                    4.62                    4.72               2.84   
min                   70.72                    0.00               0.00   
25%                   90.30                    4.44               3.26   
50%                   92.12                    8.98               5.30   
75%                   94.51                   11.48               6.67   
max                   99.06                   24.57              17.91   

       repitencia  repitencia_transicion  repitencia_primaria  \
count      482.00                 482.00               482.00   
mean         4.27                   1.45                 4.15   
std          3.40                   2.34                 3.81   
min          0.10                   0.00                 0.06   
25%          1.54                   0.22                 1.34   
50%          3.22                   0.57                 2.91   
75%          6.30                   1.54                 5.98   
max         17.37                  17.01                22.77   

       repitencia_secundaria  repitencia_media  
count                 482.00            482.00  
m

**What just happened:** The `count` row shows how many non-null values each column has. Columns where `count` is less than 482 (our total rows) have missing data. Also look at `min` and `max`: do any values look impossible? We will come back to this in Section 6.

### Inspection Summary

From our inspection, we found **5 data quality issues** to fix:

| # | Issue | Where |
|---|-------|-------|
| 1 | Missing values (NaN) | Multiple columns, especially `sedes_conectadas_a_internet`, `tamano_promedio_grupo` |
| 2 | Wrong data types | `ano` is float64, `poblacion_5_16` is object |
| 3 | Duplicate rows | 482 rows but only 462 expected (20 extra) |
| 4 | Text inconsistencies | `departamento` has too many unique values |
| 5 | Invalid values | Some percentages might be negative or above 100 |

### QUESTION 1

Based on the inspection you just ran, **list 3 specific problems** you noticed. For each one, note:
- Which column is affected
- What the problem is
- How you spotted it (which command revealed it)

*Double-click this cell and write your answer below:*

1.   valores faltantes (NaN) en varias columnas como `sedes_conectadas_a_internet`, `tamano_promedio_grupo`
2. Inconsistencia en Formatos y Mayúsculas (Ruido en Texto)
3. valores erroneos en las columnas que no tienen sentido

---

## Section 2: Missing Values

**What is NaN?** NaN stands for "Not a Number." It is pandas' way of saying "this value is missing." Think of it like reaching into your sock drawer and finding... nothing. The sock is not there. It is not zero socks (that would mean you counted and found none). It is "unknown."

**Why do missing values appear?** Data can be missing because:
- It was never collected (a survey question left blank)
- It was lost during processing (a system error)
- It does not apply (internet connectivity data before internet existed in schools)

Like socks without a pair: you need to decide whether to find a replacement, toss them, or accept the mismatch.

### 2.1 Missing Value Percentages

Before deciding what to do, we need to know how bad the problem is. Let's calculate the percentage of missing values per column.

**Run the cell below.**

In [7]:
missing_pct = (df.isnull().sum() / len(df) * 100).round(2)
missing_pct = missing_pct[missing_pct > 0].sort_values(ascending=False)

print("Missing value percentages:")
print(missing_pct)

Missing value percentages:
sedes_conectadas_a_internet    49.79
tamano_promedio_grupo          49.79
aprobacion                     11.41
reprobacion                    11.00
desercion_primaria             10.37
poblacion_5_16                 10.37
cobertura_neta_primaria        10.37
desercion                       9.75
desercion_secundaria            8.30
cobertura_neta                  8.30
departamento                    2.28
desercion_transicion            0.21
dtype: float64


### 2.2 The Decision Framework

Now we apply the decision framework from class:

| Missing % | Action | Reasoning |
|-----------|--------|----------|
| > 50% | Consider dropping the column | More gaps than data |
| < 5% | Safe to drop the rows | Losing very few rows |
| 5-50% | Fill with an appropriate value | Too many rows to lose, need to estimate |

Applying this to our columns:
- `sedes_conectadas_a_internet`, `tamano_promedio_grupo`: ~50% missing. Fill with 0 ("not reported").
- `departamento`: ~2-3% missing. Drop those rows (critical identifier, cannot guess).
- Rate columns (`desercion`, `cobertura_neta`, `aprobacion`, `reprobacion`): 8-11% missing. Fill with median.
- `poblacion_5_16`: ~10% missing. We will fix this in Section 3 (it also has type problems).

### 2.3 Fill Count-Based Columns with 0

For `sedes_conectadas_a_internet` (% schools with internet) and `tamano_promedio_grupo` (average class size), the data only exists through 2017. After that, it was not reported. Filling with 0 means "no data available for this period."

**Run the cell below.**

In [8]:
print(f"Before fillna:")
print(f"  sedes_conectadas_a_internet NaN: {df['sedes_conectadas_a_internet'].isnull().sum()}")
print(f"  tamano_promedio_grupo NaN:       {df['tamano_promedio_grupo'].isnull().sum()}")

df['sedes_conectadas_a_internet'] = df['sedes_conectadas_a_internet'].fillna(0)
df['tamano_promedio_grupo'] = df['tamano_promedio_grupo'].fillna(0)

print(f"\nAfter fillna:")
print(f"  sedes_conectadas_a_internet NaN: {df['sedes_conectadas_a_internet'].isnull().sum()}")
print(f"  tamano_promedio_grupo NaN:       {df['tamano_promedio_grupo'].isnull().sum()}")

Before fillna:
  sedes_conectadas_a_internet NaN: 240
  tamano_promedio_grupo NaN:       240

After fillna:
  sedes_conectadas_a_internet NaN: 0
  tamano_promedio_grupo NaN:       0


**What just happened:** We replaced all NaN values in those two columns with 0. The missing count went from ~240 to 0 for each.

**Why 0 is appropriate here but NOT for rates:** These columns represent counts or measurements that simply were not reported after 2017. Using 0 signals "not available." But filling a dropout rate with 0 would be misleading: 0% dropout means "nobody dropped out" (a strong claim), not "we don't know."

### 2.4 Drop Rows Where `departamento` is Missing

The `departamento` column is a critical identifier. A row without a department is like a letter without an address: useless. We cannot guess which department it belongs to, so we remove those rows.

**Run the cell below.**

In [9]:
rows_before = len(df)

df = df.dropna(subset=['departamento'])

rows_after = len(df)
print(f"Rows before: {rows_before}")
print(f"Rows after:  {rows_after}")
print(f"Removed:     {rows_before - rows_after} rows (missing departamento)")

Rows before: 482
Rows after:  471
Removed:     11 rows (missing departamento)


### 2.5 Fill Rate Columns with the Median

For columns that represent rates or percentages (dropout rate, coverage, approval, etc.), we fill with the **median**. The median is the middle value when all values are sorted. It is better than the mean because it is not distorted by extreme outliers.

**Why not fill with 0?** A 0% dropout rate means "nobody dropped out." That is very different from "we don't know." The median says: "if we had to guess, the most typical value is probably close to this."

**Run the cell below.**

In [10]:
rate_columns = [
    'tasa_matriculacion_5_16',
    'cobertura_neta', 'cobertura_neta_transicion', 'cobertura_neta_primaria',
    'cobertura_neta_secundaria', 'cobertura_neta_media',
    'cobertura_bruta', 'cobertura_bruta_transicion', 'cobertura_bruta_primaria',
    'cobertura_bruta_secundaria', 'cobertura_bruta_media',
    'desercion', 'desercion_transicion', 'desercion_primaria',
    'desercion_secundaria', 'desercion_media',
    'aprobacion', 'aprobacion_transicion', 'aprobacion_primaria',
    'aprobacion_secundaria', 'aprobacion_media',
    'reprobacion', 'reprobacion_transicion', 'reprobacion_primaria',
    'reprobacion_secundaria', 'reprobacion_media',
    'repitencia', 'repitencia_transicion', 'repitencia_primaria',
    'repitencia_secundaria', 'repitencia_media',
]

for col in rate_columns:
    n_missing = df[col].isnull().sum()
    if n_missing > 0:
        median_val = df[col].median()
        df[col] = df[col].fillna(median_val)
        print(f"{col}: filled {n_missing} NaN with median {median_val:.2f}")

print(f"\nTotal NaN remaining in rate columns: {df[rate_columns].isnull().sum().sum()}")

cobertura_neta: filled 37 NaN with median 87.95
cobertura_neta_primaria: filled 49 NaN with median 86.07
desercion: filled 46 NaN with median 3.82
desercion_transicion: filled 1 NaN with median 3.78
desercion_primaria: filled 47 NaN with median 3.15
desercion_secundaria: filled 39 NaN with median 4.75
aprobacion: filled 53 NaN with median 89.97
reprobacion: filled 53 NaN with median 6.25

Total NaN remaining in rate columns: 0


**What just happened:** Each rate column's missing values were replaced with that column's median. For example, if the median dropout rate is 3.86%, all missing dropout values now say 3.86%. This is a reasonable assumption: "if we don't know the value, assume it is typical."

### 2.6 Verify: How Many Missing Values Remain?

Let's check how our cleaning is going so far.

**Run the cell below.**

In [11]:
remaining = df.isnull().sum()
remaining = remaining[remaining > 0]

if len(remaining) == 0:
    print("No missing values remain!")
else:
    print(f"Columns still with missing values: {len(remaining)}")
    print(remaining)

Columns still with missing values: 1
poblacion_5_16    46
dtype: int64


**What to expect:** The only column with missing values should be `poblacion_5_16`. We will fix it in Section 3, because its problem is not just missing values but also wrong data types (it has commas and the text "sin dato").

---

### QUESTION 2

Why would filling `desercion` (dropout rate) with 0 be misleading? What does 0% dropout actually mean vs. "no data"?

*Double-click this cell and write your answer below:*

Llenar los espacios vacíos con 0 en la columna de desercion es un error común pero peligroso en el análisis de datos. 0% es un resultado perfecto, indicador de exito total en la retencion escolar y NaN significa ignorancia al hecho de saber que la informacion se perdio o no se reporto

---

## Section 3: Data Type Issues

Imagine sorting your laundry and finding **a shoe in the shirt pile**. It does not belong there, and it will cause problems in the wash. That is what happens when a number is stored as text: pandas cannot do math with it.

There are three main types we care about:
- `int64`: whole numbers (years, counts)
- `float64`: decimal numbers (rates, percentages)
- `object`: text/strings (names, categories)

We found two type problems:
1. `ano` (year): stored as `float64` (2023.0) instead of `int64` (2023)
2. `poblacion_5_16` (population): stored as `object` (text) instead of `int64`

### 3.1 Inspect `ano`

Let's see what the `ano` column looks like right now.

**Run the cell below.**

In [12]:
print(f"dtype: {df['ano'].dtype}")
print(f"\nSample values: {df['ano'].head(10).tolist()}")

dtype: float64

Sample values: [2023.0, 2024.0, 2016.0, 2024.0, 2020.0, 2015.0, 2022.0, 2023.0, 2011.0, 2017.0]


### 3.2 Fix `ano`: Convert to Integer

Years should be whole numbers. The CSV had mixed formats like "2011" and "2011.0", which caused pandas to read the column as float. We fix it by first ensuring all values are numeric with `pd.to_numeric()`, then converting to int.

**Run the cell below.**

In [13]:
print(f"Before: ano dtype = {df['ano'].dtype}")
print(f"Sample: {df['ano'].head(5).tolist()}")

df['ano'] = pd.to_numeric(df['ano'], errors='coerce').fillna(0).astype(int)

print(f"\nAfter: ano dtype = {df['ano'].dtype}")
print(f"Sample: {df['ano'].head(5).tolist()}")

Before: ano dtype = float64
Sample: [2023.0, 2024.0, 2016.0, 2024.0, 2020.0]

After: ano dtype = int64
Sample: [2023, 2024, 2016, 2024, 2020]


**What just happened:** The years changed from 2023.0 (float) to 2023 (integer). Clean and correct.

**The NaN-before-int trap:** If `ano` had NaN values and we tried `astype(int)` directly, pandas would crash. The `fillna(0)` step prevents that. Always fill NaN before converting to int.

### 3.3 Inspect `poblacion_5_16`

This column is text (`object`) instead of numeric. Let's look at the raw values to understand why.

**Run the cell below.**

In [16]:
print(f"Current dtype: {df['poblacion_5_16'].dtype}")
print(f"\nSample values (10 random):")
print(df['poblacion_5_16'].dropna().sample(10, random_state=42).tolist())

Current dtype: str

Sample values (10 random):
['15395', '327,302', '533666', '1288473', '22896', '161892', '544473', '487274', '393,195', '573948']


**What to notice:** Some values have commas (like "394,574") and some say "sin dato" (Spanish for "no data"). Pandas cannot convert these to numbers automatically, so it stored the whole column as text.

### 3.4 Fix `poblacion_5_16`: Clean String, Then Convert

The fix is a 2-step process:
1. **Remove the commas** with `str.replace(',', '')`
2. **Convert to numeric** with `pd.to_numeric(errors='coerce')` so "sin dato" becomes NaN instead of crashing

Then we fill the remaining NaN with 0 and convert to int.

**Run the cell below.**

In [17]:
# Step 1: Remove commas, then convert to numeric
df['poblacion_5_16'] = pd.to_numeric(
    df['poblacion_5_16'].astype(str).str.replace(',', ''),
    errors='coerce'
)

print(f"After to_numeric: dtype = {df['poblacion_5_16'].dtype}")
print(f"NaN count: {df['poblacion_5_16'].isnull().sum()}")

# Step 2: Fill NaN with 0 and convert to int
df['poblacion_5_16'] = df['poblacion_5_16'].fillna(0).astype(int)

print(f"\nFinal dtype: {df['poblacion_5_16'].dtype}")
print(f"NaN remaining: {df['poblacion_5_16'].isnull().sum()}")
print(f"Sample: {df['poblacion_5_16'].head(5).tolist()}")

After to_numeric: dtype = float64
NaN count: 78

Final dtype: int64
NaN remaining: 0
Sample: [405256, 81100, 394574, 221911, 1151149]


**What just happened:**
- Commas were removed: "394,574" became "394574"
- `pd.to_numeric()` converted valid numbers to float64
- "sin dato" and NaN values became NaN (the `errors='coerce'` flag replaces anything it cannot convert with NaN, instead of crashing)
- Finally, we filled NaN with 0 and converted to int

### YOUR TURN 1

Verify the type fixes worked. Write code to print the `dtype` and 5 sample values for both `ano` and `poblacion_5_16`.

In [6]:
import pandas as pd

df = pd.read_csv("data/educacion_estadisticas.csv")  # ← pon aquí el nombre real del archivo
print("Columna: ano")
print("dtype:", df['ano'].dtype)
print("Muestra:", df['ano'].head(5).tolist())

print("\nColumna: poblacion_5_16")
print("dtype:", df['poblacion_5_16'].dtype)
print("Muestra:", df['poblacion_5_16'].head(5).tolist())

Columna: ano
dtype: float64
Muestra: [2023.0, 2024.0, 2016.0, 2024.0, 2020.0]

Columna: poblacion_5_16
dtype: str
Muestra: ['405256', '81100', '394,574', '221911', '1151149']


---

## Section 4: Duplicates

Imagine you are folding laundry and you count the **same shirt twice**. Your count is now wrong. That is exactly what duplicate rows do: they inflate counts and distort averages. If a department appears twice for the same year with identical data, every calculation using that data is biased.

### 4.1 Count Duplicates

We use `duplicated().sum()` to count how many duplicate rows exist.

**Run the cell below.**

In [7]:
n_dupes = df.duplicated().sum()
print(f"Duplicate rows found: {n_dupes}")

Duplicate rows found: 9


### 4.2 See the Duplicates

Let's look at the actual duplicate rows. Using `keep=False` marks ALL copies (both the "original" and the "duplicate") so we can see them side by side.

**Run the cell below.**

In [8]:
dupes = df[df.duplicated(keep=False)].sort_values(['departamento', 'ano'])
print(f"Total rows involved in duplicates: {len(dupes)}")
print()
print(dupes[['ano', 'departamento', 'poblacion_5_16', 'desercion']].head(20).to_string())

Total rows involved in duplicates: 18

        ano  departamento poblacion_5_16  desercion
68   2017.0  Bogotá, D.C.        1452849       1.62
357  2017.0  Bogotá, D.C.        1452849       1.62
21   2014.0       Bolívar         490573       3.53
481  2014.0       Bolívar         490573       3.53
27   2020.0       Bolívar         466941       2.03
375  2020.0       Bolívar         466941       2.03
109  2022.0      Casanare          95609       3.98
430  2022.0      Casanare          95609       3.98
240  2015.0         Chocó         145625       4.78
255  2015.0         Chocó         145625       4.78
8    2011.0     Magdalena         333793       3.63
376  2011.0     Magdalena         333793       3.63
12   2014.0     Magdalena            NaN       2.96
117  2014.0     Magdalena            NaN       2.96
286  2016.0     Magdalena         329234       3.20
291  2016.0     Magdalena         329234       3.20
1    2024.0      Putumayo          81100       6.49
418  2024.0      Putumayo

### 4.3 Remove Duplicates

`drop_duplicates()` keeps the first occurrence of each row and removes the rest.

**Run the cell below.**

In [9]:
rows_before = len(df)

df = df.drop_duplicates()

rows_after = len(df)
print(f"Rows before: {rows_before}")
print(f"Rows after:  {rows_after}")
print(f"Removed:     {rows_before - rows_after} duplicate rows")
print(f"Duplicates remaining: {df.duplicated().sum()}")

Rows before: 482
Rows after:  473
Removed:     9 duplicate rows
Duplicates remaining: 0


**What just happened:** The duplicate rows were removed. We kept one copy of each and deleted the rest. The 20 duplicates were exact copies injected into the dataset.

### QUESTION 3

When might a duplicate row be **valid** and NOT an error? Give one example from the real world.

*Double-click this cell and write your answer below:*
Un duplicado puede ser válido cuando representa eventos distintos que comparten exactamente los mismos valores en todas las columnas.
...

---

## Section 5: Text Inconsistencies

Imagine you are folding laundry but your labels are **inconsistent**: one shirt says "Blue", another says "BLUE", another says "  blue  ". To you, they are the same color. But to pandas, these are three completely different values. This breaks any grouping or counting operation.

### 5.1 Explore the Problem

Colombia has about 34 departments (32 + Bogota D.C. + national aggregate). Let's see how many unique values our `departamento` column actually has.

**Run the cell below.**

In [10]:
print(f"Unique department values: {df['departamento'].nunique()}")
print(f"(We expect about 34)")
print(f"\nAll unique values (sorted):")
for val in sorted(df['departamento'].unique()):
    print(f"  '{val}'")

Unique department values: 122
(We expect about 34)

All unique values (sorted):


TypeError: '<' not supported between instances of 'float' and 'str'

**What just happened:** We see far more unique values than expected. The same department appears in multiple forms:
- "Antioquia", "ANTIOQUIA", "  Antioquia  ", "antioquia" are all the same department
- Some have leading/trailing spaces
- Some have accents stripped ("Narino" vs "Nariño")

To pandas, every single variation is a completely different string.

### 5.2 Standardize Text

The fix: convert everything to uppercase and remove extra whitespace with `str.upper().str.strip()`.

**Run the cell below.**

In [11]:
before_nunique = df['departamento'].nunique()

df['departamento'] = df['departamento'].str.upper().str.strip()

after_nunique = df['departamento'].nunique()

print(f"Unique values before: {before_nunique}")
print(f"Unique values after:  {after_nunique}")
print(f"Reduced by:           {before_nunique - after_nunique} values")

Unique values before: 122
Unique values after:  41
Reduced by:           81 values


**What just happened:** All department names were converted to uppercase and extra spaces were removed. The number of unique values dropped significantly. Now "Antioquia", "ANTIOQUIA", "  antioquia  " are all just "ANTIOQUIA".

### 5.3 Check for New Duplicates

Here is a subtle but important lesson: **cleaning one thing can reveal new problems.** Rows that looked different before ("Antioquia" vs "ANTIOQUIA" for the same year) are now identical after uppercasing. We need to check for duplicates again.

**Run the cell below.**

In [12]:
new_dupes = df.duplicated().sum()
print(f"New duplicates after text standardization: {new_dupes}")

if new_dupes > 0:
    rows_before = len(df)
    df = df.drop_duplicates()
    rows_after = len(df)
    print(f"Removed {rows_before - rows_after} additional duplicates")
    print(f"Final row count: {rows_after}")
else:
    print("No new duplicates created. Good.")

New duplicates after text standardization: 7
Removed 7 additional duplicates
Final row count: 466


**Lesson:** Data cleaning is iterative. Fixing one problem (text inconsistency) can create another (new duplicates). Always verify after each step.

### YOUR TURN 2

After standardization, verify how many unique departments we have now. Print the sorted list of unique department names. Does the count look reasonable for Colombia?

Write your code below.

In [ ]:
# Your code here: print nunique() and sorted list of unique departments
import pandas as pd
import unicodedata

#  Función para quitar acentos
def quitar_acentos(texto):
    if pd.isnull(texto):
        return ""
    return ''.join(c for c in unicodedata.normalize('NFD', str(texto))
                   if unicodedata.category(c) != 'Mn')

#  Convertir a string, quitar acentos, mayúsculas y espacios extra
df['departamento'] = df['departamento'].astype(str)          # todo a string
df['departamento'] = df['departamento'].apply(quitar_acentos)  # quitar tildes
df['departamento'] = df['departamento'].str.upper().str.strip() # mayúsculas y espacios

#  Diccionario para corregir nombres mal escritos
correcciones = {
    "BOGOTA, D,C,": "BOGOTA, D.C.",
    "BOGOTA, D.C": "BOGOTA, D.C.",
    "ARCHIPIELAGO DE SAN ANDRES, PROVIDENCIA Y SANTA CATALINA": 
        "ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA"
}
df['departamento'] = df['departamento'].replace(correcciones)

#  Eliminar filas duplicadas generadas por la limpieza
df = df.drop_duplicates()

# Verificación final: nunique y lista ordenada
print("Número de departamentos únicos:", df['departamento'].nunique())
print("\nLista de departamentos (ordenada):")
for dep in sorted(df['departamento'].unique()):
    print(dep)

Número de departamentos únicos: 34

Lista de departamentos (ordenada):
AMAZONAS
ANTIOQUIA
ARAUCA
ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA
ATLANTICO
BOGOTA, D.C.
BOLIVAR
BOYACA
CALDAS
CAQUETA
CASANARE
CAUCA
CESAR
CHOCO
CORDOBA
CUNDINAMARCA
GUAINIA
GUAVIARE
HUILA
LA GUAJIRA
MAGDALENA
META
NARINO
NORTE DE SANTANDER
PUTUMAYO
QUINDIO
RISARALDA
SANTANDER
SIN_DATO
SUCRE
TOLIMA
VALLE DEL CAUCA
VAUPES
VICHADA


---

## Section 6: Invalid Values

So far we have fixed missing values, wrong types, duplicates, and text inconsistencies. But there is one more problem hiding in the data: **values that exist but are impossible.**

Think of it this way: you finished your laundry, everything is folded and sorted. But then you notice a shirt labeled "Size -3" and another labeled "Size 250." Those sizes do not exist. The labels are wrong.

In our dataset, percentage columns (dropout rate, coverage, approval, etc.) must be between 0 and 100. A dropout rate of -5% is impossible. A coverage of 150% is impossible (for net coverage). These values passed all our previous checks because they are not missing, they are not the wrong type, and they are not duplicates. **They are just wrong.**

Catching these requires **domain knowledge**: knowing what valid values look like for your specific data.

### 6.1 Check Min and Max of Percentage Columns

Let's use `describe()` on just the percentage columns to check their minimum and maximum values. Any `min` below 0 or `max` above 100 is suspicious.

**Run the cell below.**

In [19]:
percentage_cols = [
    'desercion', 'desercion_primaria', 'desercion_secundaria', 'desercion_media',
    'cobertura_neta', 'cobertura_neta_primaria', 'cobertura_neta_secundaria', 'cobertura_neta_media',
    'aprobacion', 'reprobacion',
]

summary = df[percentage_cols].describe().loc[['min', 'max']].round(2)
print("Min and Max for percentage columns:")
print(summary.to_string())

Min and Max for percentage columns:
     desercion  desercion_primaria  desercion_secundaria  desercion_media  cobertura_neta  cobertura_neta_primaria  cobertura_neta_secundaria  cobertura_neta_media  aprobacion  reprobacion
min       0.46                0.34                -13.75           -10.24           -9.26                    51.12                      -5.80                  6.79       72.00       -14.42
max      10.94              118.53                 13.18           139.00          108.88                   135.06                      91.09                113.09       99.08       137.39


**What just happened:** Look at the `min` row. Do you see any negative values? Now look at the `max` row. Do you see anything above 100? Those are invalid values that should not exist in percentage columns.

### 6.2 Count the Invalid Values

Let's count exactly how many values are below 0 and above 100 in each column.

**Run the cell below.**

In [20]:
negatives = df[percentage_cols].lt(0).sum()
over_100 = df[percentage_cols].gt(100).sum()

print("Negative values (< 0) per column:")
print(negatives[negatives > 0].to_string())
print(f"\nTotal negative values: {negatives.sum()}")

print("\n" + "="*40)

print("\nValues over 100 per column:")
print(over_100[over_100 > 0].to_string())
print(f"\nTotal values over 100: {over_100.sum()}")

Negative values (< 0) per column:
desercion_secundaria         2
desercion_media              2
cobertura_neta               1
cobertura_neta_secundaria    1
reprobacion                  2

Total negative values: 8


Values over 100 per column:
desercion_primaria          1
desercion_media             1
cobertura_neta             11
cobertura_neta_primaria     9
cobertura_neta_media        1
reprobacion                 1

Total values over 100: 24


### 6.3 See the Invalid Rows

Let's look at the actual rows with invalid values so we can understand the scope of the problem.

**Run the cell below.**

In [21]:
# Find rows with any negative percentage value
mask_negative = (df[percentage_cols] < 0).any(axis=1)
mask_over_100 = (df[percentage_cols] > 100).any(axis=1)

print(f"Rows with negative values: {mask_negative.sum()}")
if mask_negative.sum() > 0:
    print(df.loc[mask_negative, ['ano', 'departamento'] + percentage_cols].to_string())

print(f"\nRows with values > 100: {mask_over_100.sum()}")
if mask_over_100.sum() > 0:
    print(df.loc[mask_over_100, ['ano', 'departamento'] + percentage_cols].to_string())

Rows with negative values: 8
        ano departamento  desercion  desercion_primaria  desercion_secundaria  desercion_media  cobertura_neta  cobertura_neta_primaria  cobertura_neta_secundaria  cobertura_neta_media  aprobacion  reprobacion
33   2013.0        HUILA       4.78                3.27                  7.00             4.99           -9.26                    90.86                      75.27                 39.07       92.18         3.04
44   2011.0      QUINDIO       5.04                4.52                  5.98             4.10           90.83                    85.53                      -5.80                 47.29       92.43         2.53
139  2011.0       CALDAS       4.13                3.92                -13.75             2.82           86.98                    83.86                      73.41                 45.38       90.62         5.25
153  2016.0        CAUCA       2.84                 NaN                  4.25           -10.00           83.84                    8

### 6.4 Fix Invalid Values

We will replace invalid values with NaN, then fill them with the column median (the same strategy we used for missing values). This is the safer approach: we treat impossible values the same as missing data.

**Run the cell below.**

In [23]:
import numpy as np  # <<--- importante
import pandas as pd

total_fixed = 0

for col in percentage_cols:
    # Contar valores inválidos (<0 o >100)
    invalid_mask = (df[col] < 0) | (df[col] > 100)
    n_invalid = invalid_mask.sum()
    
    if n_invalid > 0:
        # Reemplazar inválidos con NaN
        df.loc[invalid_mask, col] = np.nan
        
        # Rellenar NaN con la mediana
        median_val = df[col].median()
        df[col] = df[col].fillna(median_val)
        
        print(f"{col}: fixed {n_invalid} invalid values (replaced with median {median_val:.2f})")
        total_fixed += n_invalid

print(f"\nTotal invalid values fixed: {total_fixed}")

desercion_primaria: fixed 1 invalid values (replaced with median 3.13)
desercion_secundaria: fixed 2 invalid values (replaced with median 4.75)
desercion_media: fixed 3 invalid values (replaced with median 3.28)
cobertura_neta: fixed 12 invalid values (replaced with median 87.50)
cobertura_neta_primaria: fixed 9 invalid values (replaced with median 85.96)
cobertura_neta_secundaria: fixed 1 invalid values (replaced with median 72.73)
cobertura_neta_media: fixed 1 invalid values (replaced with median 41.20)
reprobacion: fixed 3 invalid values (replaced with median 6.26)

Total invalid values fixed: 32


**What just happened:** We found values that were mathematically present but logically impossible (negative percentages, percentages above 100). We replaced them with NaN and then filled with the median, just like we do with missing values.

**Key takeaway:** Domain knowledge is essential for data cleaning. Without knowing that dropout rates must be between 0 and 100, we would never catch these errors. The `describe()` function is your friend here: always check `min` and `max` against what you know about the data.

### 6.5 Verify the Fix

Let's confirm that all percentage values are now within the valid 0-100 range.

**Run the cell below.**

In [24]:
print("After fix - Min and Max for percentage columns:")
print(df[percentage_cols].describe().loc[['min', 'max']].round(2).to_string())

remaining_invalid = (df[percentage_cols].lt(0).sum().sum() + df[percentage_cols].gt(100).sum().sum())
print(f"\nInvalid values remaining: {remaining_invalid}")

After fix - Min and Max for percentage columns:
     desercion  desercion_primaria  desercion_secundaria  desercion_media  cobertura_neta  cobertura_neta_primaria  cobertura_neta_secundaria  cobertura_neta_media  aprobacion  reprobacion
min       0.46                0.34                  0.38             0.42           50.79                    51.12                      24.06                  6.79       72.00          0.0
max      10.94               10.73                 13.18            29.89           99.52                    99.47                      91.09                 65.32       99.08         17.6

Invalid values remaining: 0


### QUESTION 4

Which approach is better for invalid percentage values:

**(a)** Replace with NaN and fill with the median (what we did), or  
df[percentage_cols].describe().loc[['min', 'max']]

**(b)** Clip to the valid range (set negatives to 0 and values >100 to 100)?

remaining_invalid = (df[percentage_cols].lt(0).sum().sum() + df[percentage_cols].gt(100).sum().sum())
Think about what each approach assumes. When might clipping be better? When might the median approach be better?

*Double-click this cell and write your answer below:*

...

In [31]:
df[percentage_cols].describe().loc[['min', 'max']]


desercion  desercion_primaria  desercion_secundaria  desercion_media  \
min       0.46                0.34                  0.38             0.42   
max      10.94               10.73                 13.18            29.89   

     cobertura_neta  cobertura_neta_primaria  cobertura_neta_secundaria  \
min           50.79                    51.12                      24.06   
max           99.52                    99.47                      91.09   

     cobertura_neta_media  aprobacion  reprobacion  
min                  6.79       72.00          0.0  
max                 65.32       99.08         17.6

In [32]:
remaining_invalid = (df[percentage_cols].lt(0).sum().sum() + df[percentage_cols].gt(100).sum().sum())

---

## Section 7: Summary

We have completed all 5 cleaning steps. Let's compare the original dataset with our cleaned version to see the full impact of our work.

**Run the cell below.**

In [36]:
# Guardar copia del dataframe original antes de limpiar

print("=" * 55)
print("  BEFORE CLEANING (original)")
print("=" * 55)
print(f"  Rows:                {len(df_original)}")
print(f"  Total NaN:           {df_original.isnull().sum().sum()}")
print(f"  Duplicates:          {df_original.duplicated().sum()}")
print(f"  Unique departments:  {df_original['departamento'].nunique()}")
print(f"  ano dtype:           {df_original['ano'].dtype}")
print(f"  poblacion dtype:     {df_original['poblacion_5_16'].dtype}")

print()

print("=" * 55)
print("  AFTER CLEANING")
print("=" * 55)
print(f"  Rows:                {len(df)}")
print(f"  Total NaN:           {df.isnull().sum().sum()}")
print(f"  Duplicates:          {df.duplicated().sum()}")
print(f"  Unique departments:  {df['departamento'].nunique()}")
print(f"  ano dtype:           {df['ano'].dtype}")
print(f"  poblacion dtype:     {df['poblacion_5_16'].dtype}")

print()
print("=" * 55)
print("  INVALID VALUES CHECK")
print("=" * 55)
pct_cols_check = ['desercion', 'cobertura_neta', 'aprobacion', 'reprobacion']
for col in pct_cols_check:
    print(f"  {col}: min={df[col].min():.2f}, max={df[col].max():.2f}")

  BEFORE CLEANING (original)
  Rows:                466
  Total NaN:           616
  Duplicates:          0
  Unique departments:  34
  ano dtype:           float64
  poblacion dtype:     str

  AFTER CLEANING
  Rows:                466
  Total NaN:           616
  Duplicates:          0
  Unique departments:  34
  ano dtype:           float64
  poblacion dtype:     str

  INVALID VALUES CHECK
  desercion: min=0.46, max=10.94
  cobertura_neta: min=50.79, max=99.52
  aprobacion: min=72.00, max=99.08
  reprobacion: min=0.00, max=17.60


### The 5-Step Cleaning Workflow

Here is the workflow you just practiced. Use these exact steps in every data project:

| Step | What | Key Commands |
|------|------|--------------|
| 1. **Inspect** | Understand the data before touching it | `df.shape`, `df.dtypes`, `df.isnull().sum()`, `df.describe()` |
| 2. **Handle missing** | Decide: drop, fill with 0, or fill with median | `fillna(0)`, `fillna(median)`, `dropna(subset=...)` |
| 3. **Fix types** | Numbers stored as text, floats that should be int | `pd.to_numeric(errors='coerce')`, `astype(int)` |
| 4. **Remove duplicates** | Exact copies that inflate counts | `duplicated().sum()`, `drop_duplicates()` |
| 5. **Validate values** | Impossible values based on domain knowledge | `describe()` min/max, boolean masks |

**Remember:** Cleaning is iterative. Fixing one problem can create another (like text standardization creating new duplicates). Always verify after each step.

This dataset is now ready for analysis.

### FINAL REFLECTION

Write 2-3 sentences:

1. What was the most surprising thing you found during cleaning?
2. Which of the 5 steps do you think is most important? Why?
3. How would you apply these steps to your project dataset from datos.gov.co?

Lo más sorprendente fue encontrar valores inválidos, como porcentajes negativos o mayores a 100, que no se detectan automáticamente y podrían distorsionar cualquier análisis. Considero que el paso más importante es validar los valores, porque incluso si no hay datos faltantes ni duplicados, el dataset puede contener errores lógicos o imposibles

...